### ISSUE TITLE
Complete crash with userns-remap option

### ISSUE THREAD

**Issue by Soullivaneuh** [ubuntu/microk8s#337]
In order to have only one docker on my local machine even for non-kubernetes projects, I aliased the microk8s one and followed [this solution]() to avoid root file creation:

```
➜ cat /var/snap/microk8s/current/args/dockerd
--add-runtime nvidia=${SNAP}/usr/bin/nvidia-container-runtime
-H unix://${SNAP_DATA}/docker.sock
--exec-root ${SNAP_COMMON}/var/run/docker
--graph ${SNAP_COMMON}/var/lib/docker
--pidfile ${SNAP_COMMON}/docker-pid
--config-file=${SNAP_DATA}/args/docker-daemon.json
--userns-remap=sullivan
```

But with this option, none of the containers in kubernetes want to start:

```
➜ kubectl get all --all-namespaces
NAMESPACE            NAME                                                  READY   STATUS             RESTARTS   AGE
container-registry   pod/registry-7fc4594d64-9x5jt                         0/1     CrashLoopBackOff   11         25m
kube-system          pod/heapster-v1.5.2-64874f6bc6-lcrdk                  0/4     CrashLoopBackOff   44         16m
kube-system          pod/hostpath-provisioner-599db8d5fb-h5l7p             0/1     CrashLoopBackOff   23         25m
kube-system          pod/kube-dns-6ccd496668-4g4h5                         0/3     CrashLoopBackOff   75         25m
kube-system          pod/kubernetes-dashboard-654cfb4879-tn666             0/1     CrashLoopBackOff   22         25m
kube-system          pod/monitoring-influxdb-grafana-v4-6679c46745-v2pjr   0/2     CrashLoopBackOff   44         25m
kube-system          pod/tiller-deploy-69ffbf64bc-55lmw                    0/1     CrashLoopBackOff   22         25m

NAMESPACE            NAME                           TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)             AGE
container-registry   service/registry               NodePort    10.152.183.63    <none>        5000:32000/TCP      25m
default              service/kubernetes             ClusterIP   10.152.183.1     <none>        443/TCP             26m
kube-system          service/heapster               ClusterIP   10.152.183.57    <none>        80/TCP              25m
kube-system          service/kube-dns               ClusterIP   10.152.183.10    <none>        53/UDP,53/TCP       25m
kube-system          service/kubernetes-dashboard   ClusterIP   10.152.183.116   <none>        443/TCP             25m
kube-system          service/monitoring-grafana     ClusterIP   10.152.183.141   <none>        80/TCP              25m
kube-system          service/monitoring-influxdb    ClusterIP   10.152.183.180   <none>        8083/TCP,8086/TCP   25m
kube-system          service/tiller-deploy          ClusterIP   10.152.183.124   <none>        44134/TCP           25m

NAMESPACE            NAME                                             READY   UP-TO-DATE   AVAILABLE   AGE
container-registry   deployment.apps/registry                         0/1     1            0           25m
kube-system          deployment.apps/heapster-v1.5.2                  0/1     1            0           25m
kube-system          deployment.apps/hostpath-provisioner             0/1     1            0           25m
kube-system          deployment.apps/kube-dns                         0/1     1            0           25m
kube-system          deployment.apps/kubernetes-dashboard             0/1     1            0           25m
kube-system          deployment.apps/monitoring-influxdb-grafana-v4   0/1     1            0           25m
kube-system          deployment.apps/tiller-deploy                    0/1     1            0           25m

NAMESPACE            NAME                                                        DESIRED   CURRENT   READY   AGE
container-registry   replicaset.apps/registry-7fc4594d64                         1         1         0       25m
kube-system          replicaset.apps/heapster-v1.5.2-56c546dbb8                  0         0         0       16m
kube-system          replicaset.apps/heapster-v1.5.2-64874f6bc6                  1         1         0       16m
kube-system          replicaset.apps/heapster-v1.5.2-6bc7c4965d                  0         0         0       25m
kube-system          replicaset.apps/hostpath-provisioner-599db8d5fb             1         1         0       25m
kube-system          replicaset.apps/kube-dns-6ccd496668                         1         1         0       25m
kube-system          replicaset.apps/kubernetes-dashboard-654cfb4879             1         1         0       25m
kube-system          replicaset.apps/monitoring-influxdb-grafana-v4-6679c46745   1         1         0       25m
kube-system          replicaset.apps/tiller-deploy-69ffbf64bc                    1         1         0       25m
```

And I have no log from them.

If I remove the `userns-remap`, it work as indented.

Note: A simple `docker run` works, the issue is only with Kubernetes.

It looks like a bug, but I have currently no idea how to troubleshoot this...

[inspection-report-20190225_211409.tar.gz](https://github.com/ubuntu/microk8s/files/2902444/inspection-report-20190225_211409.tar.gz)

**Comment by balchua** [comment]
I am not sure if this is doable in k8s even as of 1.14.  there is an issue open and i dont think it will land in 1.14.

https://github.com/kubernetes/enhancements/issues/127

Did u get this user namespace remapping in none microk8s, like minikube?

**Comment by ktsakalozos** [comment]
@Soullivaneuh you may have heard we will be switching to containerd with the 1.14 release. This should address the problems we have when deploying MicroK8s next to an already existing dockerd. You can try a preview of MicroK8s with containerd with:
```
snap install microk8s --classic --channel=1.13/edge/secure-containerd
```

Let me know how we can smooth this transition for you. Thanks.

**Comment by Soullivaneuh** [comment]
> when deploying MicroK8s next to an already existing dockerd

I don't have any other dockerd, I expecially want to use the microk8s one only. :-)

**Comment by Soullivaneuh** [comment]
Oh right @ktsakalozos, I misunderstood what you said.

So I will not be able to use docker directly on microk8s.

What should I do for classic docker-composed legacy project? Can I still interact them with microk8s in some ways?

**Comment by ktsakalozos** [comment]
Dockerd will not be available in the 1.14 release of MicroK8s; dockerd is going to be replaced with containerd.

If you want dockerd from MicroK8s you should stick to the 1.13 release series with:
```
snap refresh --channel=1.13/stable microk8s
```

More info on what is coming up in v1.14: https://itnext.io/containerd-on-a-more-secure-microk8s-c2c97ac0268f

If you want to use docker-compose you should deploy docker with `apt-get install docker.io` and use it as you normally would. Have you seen the Kompose project https://github.com/kubernetes/kompose ? This should be able to create the yamls for Kubernetes in the same way docker-compose does.

**Comment by Soullivaneuh** [comment]
> If you want dockerd from MicroK8s you should stick to the 1.13 release

Hell no, let's move on! :-)

I'm sure you know what you are doing. ;-)

I'll just re-install my own docker in that case and yes, I know kompose, I'm learning how to use it right now.

Thanks for the great explanation! :+1:
